src/main/java/org/candlepin/model/ConsumerCurator.java

```diff
@@ -1001,6 +1008,107 @@ else if (systemUuid != null) {
         return found;
     }
 
+    /**
+     * Deletes inactive consumers and creates {@link DeletedConsumer} records.
+     * The identity certificate and content access certificate for the consumers
+     * are removed and their serials are revoked.
+     *
+     * @param inactiveConsumerIds - ids of inactive consumers to delete.
+     * @return the number of consumers that have been deleted.
+     */
+    @Transactional
```

Comment:
We should avoid adding transaction management to/in curator methods.

Reply:
Removed, and added the transaction management in the InactiveConsumerCleanerJob

Reply:
So, after writing a spec test to test the job, I was running into an error about the update needing to be in a transaction when the method in InactiveConsumerCleanerJob that called the curator had the transactional annotation. Moving the Transactional annotation back to the curator method resolved that error.

Reply:
Let's identify the root cause behind that error since there must be something else going wrong; we should be able to set the transaction level at the level we want

Reply:
What error? In general, if a given function only works because of the transactional annotation, that means our transaction management is busted, or a utility method has a hard-coded flush that needs to be gated or removed. We should seek to fix that issue rather than patching it up in a way that obscures/bandaids it as we've done in the past.

Our transaction management on the whole is pretty bad as a result of misuse of these annotations, and this has a direct impact on our locking model. We should avoid creating new situations where the boundaries of our transactions are unclear if at all possible.

Reply:
So, instead of using the Transactional annotation on a method in InactiveConsumerCleanerJob that calls the delete method in the curator, I used the org.candlepin.util.Transactional wrapper and that worked!

```
      Transactional<Integer> transaction = this.consumerCurator.transactional((args) -> {
          if (args == null || args.length != 1) {
              throw new IllegalArgumentException("Unexpected value received for arguments: " + args);
          }
      
          List<String> consumerIds = (List<String>) args[0];
          return consumerCurator.deleteInactiveConsumers(consumerIds);
      });
```

Reply:
Hrmm... looks like any bulk update or delete op at the Hibernate/JPA layer requires a transaction. That's a bit unfortunate/annoying, as it kinda violates our general design goals here. In the case where a curator method is doing a flush and the failure is a result of that, that's definitely something we should be dealing with via better transaction management or moving flush control out. But this case, I'm uncertain. I would like our curator methods to be callable from any context when they're not doing weird things.

Thoughts?

Reply:
@Ceiu  I see what you are saying! I almost think that we need to break up the deleteInactiveConsumers method in the ConsumerCurator class into individual methods that do one thing, and for the bulk update/delete operations we just build the Query object and return it. Then have a deleteInactiveConsumers method in a Manager (org.candlepin.controller package) class that handles the transaction and calling each individual curator method and invoking the updates/deletes.

Reply:
I'm not sure I understand the problem. Wasn't the delete in a transaction before anyway, since the @Transactional annotation was on it? What is the difference with using the annotation and the transactional block in this case?

Also, I would agree that conceptually the deleteInactiveConsumers method shouldn't be in a curator, but in the job class itself, which then calls into the individual curator methods (the deletedConsumersStatement query would have to be extracted into it's own curator method in that case)

Reply:
Agreed. I'm not too fond of spinning up a controller/manager class for one-offs, but the method could definitely be paired down as @nikosmoum described. The curators are intended to be minimalist classes which perform database operations, with any surrounding logic being limited to data shaping. We don't always get it right, but that's the goal.

As far as the exact problem, I'm not *entirely* sure -- looks like it could be a new restriction as of Hibernate 5.3? I took the exception Josh hit and did some googling. Many of the results don't really go over the why, and are just people telling others to slap the annotation on the method in question and move on; which pains me more than I can express in words. I didn't see anything clearly stated in the Hibernate user guide regarding bulk operations requiring a transaction, so there might be something else at play here worth investigating.